src/Orleans.Transactions/DeadlockDetection/DeadlockDetector.cs

```diff
@@ -70,27 +70,28 @@ public Batch(IEnumerable<SiloAddress> silos)
         private int MaxRequestCount { get; } = 5;
         private TimeSpan RequestTimeout { get; } = TimeSpan.FromSeconds(2);
 
-        private readonly ISiloStatusOracle siloStatusOracle;
+        private ISiloStatusOracle siloStatusOracle;
         private readonly ILogger<DeadlockDetector> logger;
-        private readonly ITransactionalLockObserver localObserver;
         private readonly IDictionary<Guid, Batch> batches =
             new Dictionary<Guid, Batch>();
 
-        private readonly IInternalGrainFactory internalGrainFactory;
+        private IInternalGrainFactory internalGrainFactory;
 
-        internal DeadlockDetector(ILogger<DeadlockDetector> logger,
-            ITransactionalLockObserver localObserver,
-            ISiloStatusOracle siloStatusOracle,
-            IInternalGrainFactory internalGrainFactory)
+        public DeadlockDetector(ILogger<DeadlockDetector> logger)
         {
             this.logger = logger;
-            this.localObserver = localObserver;
-            this.siloStatusOracle = siloStatusOracle;
-            this.internalGrainFactory = internalGrainFactory;
+        }
+
+        public override async Task OnActivateAsync()
+        {
+            await base.OnActivateAsync();
+            this.siloStatusOracle = this.ServiceProvider.GetRequiredService<ISiloStatusOracle>();
+            this.internalGrainFactory = this.ServiceProvider.GetRequiredService<IInternalGrainFactory>();
```

Comment:
Is this change for testing purposes, or some other reason?

Reply:
You can't access these services via DI.